src/Viz.js

```diff
@@ -310,39 +310,60 @@ export default class Viz {
+        this._pointShader = null;
+        this._lineShader = null;
+        this._polygonShader = null;
+    }
 
-        if (!this._geomType || this._geomType === 'point') {
-            this.pointShader = compileShader(gl,
-                { vertexShader: pointVertexShaderGLSL, fragmentShader: pointFragmentShaderGLSL },
-                { transform: this.transform }, this);
-        }
-        if (!this._geomType || this._geomType === 'line') {
-            this.lineShader = compileShader(gl,
-                { vertexShader: lineVertexShaderGLSL, fragmentShader: lineFragmentShaderGLSL },
-                { transform: this.transform }, this);
-        }
-        if (!this._geomType || this._geomType === 'polygon') {
-            this.polygonShader = compileShader(gl,
-                { vertexShader: polygonVertexShaderGLSL, fragmentShader: polygonFragmentShaderGLSL },
-                { transform: this.transform }, this);
+    get colorShader () {
+        return this._compileShader('colorShader', shaders.styler.colorShaderGLSL, { color: this.color });
+    }
+    get widthShader () {
+        return this._compileShader('widthShader', shaders.styler.widthShaderGLSL, { width: this.width });
+    }
+    get strokeColorShader () {
+        return this._compileShader('strokeColorShader', shaders.styler.colorShaderGLSL, { color: this.strokeColor });
+    }
+    get strokeWidthShader () {
+        return this._compileShader('strokeWidthShader', shaders.styler.widthShaderGLSL, { width: this.strokeWidth });
+    }
+    get filterShader () {
+        return this._compileShader('filterShader', shaders.styler.filterShaderGLSL, { filter: this.filter });
+    }
+    get symbolShader () {
+        return this._compileShader('symbolShader', shaders.symbolizer.symbolShaderGLSL, {
+            symbol: this.symbol,
+            symbolPlacement: this.symbolPlacement,
+            transform: this.transform
+        });
+    }
+    get pointShader () {
+        return this._compileShader('pointShader', { vertexShader: pointVertexShaderGLSL, fragmentShader: pointFragmentShaderGLSL },
+            { transform: this.transform });
+    }
+    get lineShader () {
+        return this._compileShader('lineShader', { vertexShader: lineVertexShaderGLSL, fragmentShader: lineFragmentShaderGLSL },
+            { transform: this.transform });
+    }
+    get polygonShader () {
+        return this._compileShader('polygonShader', { vertexShader: polygonVertexShaderGLSL, fragmentShader: polygonFragmentShaderGLSL },
+            { transform: this.transform });
+    }
+
+    _compileShader (shaderName, GLSL, expr) {
+        if (!this['_' + shaderName]) {
+            this._getRootExpressions().forEach(expr => expr._bindMetadata(this.metadata));
+            checkVizPropertyTypes(this);
+            this['_' + shaderName] = compileShader(this.gl, GLSL, expr, this);
         }
+        return this['_' + shaderName];
     }
 
     replaceChild (toReplace, replacer) {
```

Comment:
This could be refactored to something like (I haven't tested it, though):
```suggestion
	replaceChild (toReplace, replacer) {
        if (Object.values(this.variables).includes(toReplace)) {
            const varName = Object.keys(this.variables).find(varName => this.variables[varName] === toReplace);
            this.variables[varName] = replacer;
            replacer.parent = this;
            replacer.notify = toReplace.notify;
        } else {
        	const properties = ['color', 'width', 'strokeColor', 'strokeWidth', 'filter', 'symbol', 'symbolPlacement', 'transform'];
            properties.forEach((property) => {
            	if (toReplace === this[property]){
                	this[property] = replacer;
                    replacer.parent = this;
                    replacer.notify = toReplace.notify;
                }
            });
            
            if (!properties.includes(toReplace)){
            	throw new CartoRuntimeError('No child found');
            }
        }
  	}
```

Reply:
Yes! :+1: